common/util/vector_tree.h

```diff
@@ -198,6 +200,172 @@ class _VectorTreeImpl {
+// parent pointer in removed nodes.
+template <class Container>
+class VectorTreeChildrenList
+    : ContainerProxyBase<VectorTreeChildrenList<Container>, Container> {
+  using Base = ContainerProxyBase<VectorTreeChildrenList<Container>, Container>;
+  friend Base;
+
+  using ThisType = VectorTreeChildrenList<Container>;
+
+  using VectorTreeType = typename Container::value_type;
+  friend VectorTreeType;
+
+ public:
+  using typename Base::container_type;
+
+  // Sequence Container interface
+
+  using typename Base::value_type;
+
+  using typename Base::const_reference;
+  using typename Base::reference;
+
+  using typename Base::const_iterator;
+  using typename Base::iterator;
+
+  using typename Base::difference_type;
+  using typename Base::size_type;
+
+  using typename Base::const_reverse_iterator;
+  using typename Base::reverse_iterator;
+
+  using Base::begin;
+  using Base::cbegin;
+  using Base::cend;
+  using Base::end;
+
+  using Base::crbegin;
+  using Base::crend;
+  using Base::rbegin;
+  using Base::rend;
+
+  using Base::back;
+  using Base::front;
+  using Base::operator[];
+  using Base::at;
+
+  using Base::empty;
+  using Base::max_size;
+  using Base::size;
+
+  using Base::emplace_back;
+  using Base::push_back;
+
+  using Base::emplace_front;
+  using Base::push_front;
+
+  using Base::emplace;
+  using Base::insert;
+
+  using Base::clear;
+  using Base::erase;
+  using Base::pop_back;
+  using Base::pop_front;
+
+  using Base::assign;
+  using Base::operator=;
+  using Base::swap;
+
+  using Base::capacity;
+  using Base::reserve;
+  using Base::resize;
+
+ protected:
+  // ContainerProxy interface
+
+  container_type& underlying_container() { return container_; }
+  const container_type& underlying_container() const { return container_; }
+
+  void ElementsInserted(iterator first, iterator last) {
+    for (auto& child : iterator_range(first, last)) {
+      child.parent_ = &node_;
+    }
+  }
+
+  // Unused:
+  // void ElementsBeingRemoved(iterator first, iterator last)
+
+  // Unused:
+  // void ElementsBeingReplaced()
+
+  void ElementsWereReplaced() {
+    for (auto& child : container_) {
+      child.parent_ = &node_;
+    }
+  }
+
+ private:
+  // Hide constructors and assignments from the world. This object is created
+  // and assigned-to only in VectorTree.
+
+  explicit VectorTreeChildrenList(VectorTreeType& node) : node_(node) {}
+
+  // Construction requires parent node reference.
+  VectorTreeChildrenList(const VectorTreeChildrenList&) = delete;
+
+  VectorTreeChildrenList(VectorTreeType& node,
+                         const VectorTreeChildrenList& other)
+      : node_(node), container_(other.container_) {
+    for (auto& child : container_) {
+      child.parent_ = &node;
+    }
+  }
+
+  VectorTreeChildrenList& operator=(const VectorTreeChildrenList& other) {
+    container_ = other.container_;
+    for (auto& child : container_) {
```

Comment:
Call one of the above helper functions.
(Same elsewhere below.)

Reply:
Calling `LinkChildrenToParent` now.